### PR TITLE
Seekable source

### PR DESCRIPTION
**Pull request by varsill**

### CODE REVIEW

File: lib/membrane_file/source.ex [hunk 1]
```diff
@@ -57,19 +102,45 @@ defmodule Membrane.File.Source do
...
+      buffer_actions ++
+        if state.should_send_eos? and (state.size_to_read == 0 or supplied_size < to_supply_size) do
+          [end_of_stream: :output]
+        else
+          if to_supply_size == supplied_size do
+            redemand
+          else
+            []
+          end
+        end
```
Comment (mat-hek) [inline lib/membrane_file/source.ex:142]: convert to `cond`

**Review by mat-hek — CHANGES_REQUESTED** [review]

File: lib/membrane_file/seek_source_event.ex [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  data from the given position in file, specified by the `:start` field of the event's
+  struct. The source reads up to `size_to_read` bytes of the data from file (it can
+  read less if the file ends).
+  If the event is set with `last?: true`, once `size_to_read` bytes are read or the
+  file ends, the source will return `end_of_stream` action on the `:output` pad.
+  """
+  @derive Membrane.EventProtocol
+
+  @type offset_t :: integer()
+  @type position_t :: offset_t() | {:bof | :cur | :eof, offset_t()} | :bof | :cur | :eof
```
[the remark this answers is not in the document]
Comment (varsill) [reply lib/membrane_file/seek_source_event.ex:16]: In fact it's not my invention, it's specified by: https://www.erlang.org/doc/man/file.html#position-2
We also use it here: https://github.com/membraneframework/membrane_file_plugin/blob/7319d702549d6ddb56dd6137dd65a01acf490bfe/lib/membrane_file/seek_event.ex#L18 where and we don't describe it there.
Surely I can add some reference link to the documentation in both these places :D

File: lib/membrane_file/source.ex [hunk 3]
```diff
@@ -2,10 +2,19 @@ defmodule Membrane.File.Source do
...
   Element that reads chunks of data from given file and sends them as buffers
   through the output pad.
+
+  Can work in two modes, determined by the `seekable?` option.
+  With `seekable?: false`, the source will start reading data from the file exactly the moment it starts
+  playing and will read it till the end of file, setting the `end_of_stream` action on the `:output` pad
+  when the reading is done.
+  With `seekable?: true`, the process of reading is driven by receiving `Membrane.File.SeekSourceEvent` events.
+  The source working in `seekable?: true` mode won't send any data before that event is received.
+  For more information about how to steer reading in `seekable?: true` mode, see: `Membrane.File.SeekSourceEvent`.
```
[the remark this answers is not in the document]
Comment (varsill) [reply lib/membrane_file/source.ex:12]: done

File: lib/membrane_file/source.ex [hunk 4]
```diff
@@ -43,6 +65,29 @@ defmodule Membrane.File.Source do
...
+  @impl true
+  def handle_event(
+        :output,
+        %SeekSourceEvent{start: seek_start, size_to_read: size_to_read, last?: last?},
+        _ctx,
+        %{seekable?: true} = state
+      ) do
+    @common_file.seek!(state.fd, seek_start)
+
+    {[event: {:output, %Membrane.File.NewSeekEvent{}}, redemand: :output],
```
[the remark this answers is not in the document]
Comment (varsill) [reply lib/membrane_file/source.ex:77]: done

File: lib/membrane_file/source.ex (hunk 1 above)
Comment (varsill) [reply lib/membrane_file/source.ex:142]: done

**Review by mat-hek — APPROVED** [review]

**Pull request merged**